translate/v34tov33/v34tov33.go

```diff
@@ -0,0 +1,135 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v34tov33
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/ignition/config/util"
+	"github.com/coreos/ignition/v2/config/translate"
+	"github.com/coreos/ignition/v2/config/v3_3/types"
+	old_types "github.com/coreos/ignition/v2/config/v3_4/types"
+	"github.com/coreos/ignition/v2/config/validate"
+)
+
+// Copy of github.com/coreos/ignition/config/v3_4/translate/translate.go
+// with the types & old_types imports reversed (the referenced file translates
+// from 3.3 -> 3.4 but as a result only touches fields that are understood by
+// the 3.3 spec).
+
+func translateIgnition(old old_types.Ignition) (ret types.Ignition) {
+	// use a new translator so we don't recurse infinitely
+	translate.NewTranslator().Translate(&old, &ret)
+	ret.Version = types.MaxVersion.String()
+	return
+}
+
+func translateFileEmbedded1(old old_types.FileEmbedded1) (ret types.FileEmbedded1) {
+	tr := translate.NewTranslator()
+	tr.Translate(&old.Append, &ret.Append)
+	tr.Translate(&old.Contents, &ret.Contents)
+	if old.Mode != nil {
+		ret.Mode = util.IntToPtr(*old.Mode)
+	}
+	return
+}
+
+func translateDirectoryEmbedded1(old old_types.DirectoryEmbedded1) (ret types.DirectoryEmbedded1) {
+	if old.Mode != nil {
+		ret.Mode = util.IntToPtr(*old.Mode)
+	}
+	return
+}
+
+func translateLuks(old old_types.Luks) (ret types.Luks) {
+	tr := translate.NewTranslator()
+	tr.AddCustomTranslator(translateTang)
+	tr.Translate(&old.Clevis, &ret.Clevis)
+	tr.Translate(&old.Device, &ret.Device)
+	tr.Translate(&old.KeyFile, &ret.KeyFile)
+	tr.Translate(&old.Label, &ret.Label)
+	tr.Translate(&old.Name, &ret.Name)
+	tr.Translate(&old.Options, &ret.Options)
+	tr.Translate(&old.UUID, &ret.UUID)
+	tr.Translate(&old.WipeVolume, &ret.WipeVolume)
+	return
+}
+
+func translateTang(old old_types.Tang) (ret types.Tang) {
+	tr := translate.NewTranslator()
+	tr.Translate(&old.Thumbprint, &ret.Thumbprint)
+	tr.Translate(&old.URL, &ret.URL)
+	return
+}
+
+func translateConfig(old old_types.Config) (ret types.Config) {
+	tr := translate.NewTranslator()
+	tr.AddCustomTranslator(translateIgnition)
+	tr.AddCustomTranslator(translateDirectoryEmbedded1)
+	tr.AddCustomTranslator(translateFileEmbedded1)
+	tr.AddCustomTranslator(translateLuks)
+	tr.Translate(&old, &ret)
+	return
+}
+
+// end copied Ignition v3_4/translate block
+
+// Translate translates Ignition spec config v3.4 to spec v3.3
+func Translate(cfg old_types.Config) (types.Config, error) {
+	rpt := validate.ValidateWithContext(cfg, nil)
+	if rpt.IsFatal() {
+		return types.Config{}, fmt.Errorf("Invalid input config:\n%s", rpt.String())
+	}
+
+	// 3.3 does not support tang offline provisioning
+	for _, luks := range cfg.Storage.Luks {
+		for _, tang := range luks.Clevis.Tang {
+			if tang.Advertisement != nil && *tang.Advertisement != "" {
```

Comment:
Please use the `util` helper.

Reply:
hmm, I did look at the util helper but for some reason it did not have the string helpers I expected. I will take another look.

Reply:
We're importing the regular Ignition one.

Reply:
Ah I see, it was importing  "github.com/coreos/ignition/config/util"; but we want "github.com/coreos/ignition/v2/config/util"